api/src/wfl/api/handlers.clj

```diff
@@ -118,65 +112,72 @@
-               (fail-with-response response/not-found (format "Workload %s not found" uuid))
-               (succeed result))))
-         (success []))))))
+  (letfn [(go! [tx {:keys [uuid id] :as workload}]
+            (if-not (:finished workload)
+              (do
+                (logr/infof "updating workload %s" uuid)
+                (postgres/update-workload! tx workload)
+                (postgres/load-workload-for-id tx id))
+              workload))]
+    (log-and-fail-on-error
+      (jdbc/with-db-transaction [tx (postgres/wfl-db-config)]
+        (succeed
+          (mapv (partial go! tx)
+            (if-let [uuid (-> request :parameters :query :uuid)]
+              (if-let [workload (postgres/load-workload-for-uuid tx uuid)]
+                [workload]
+                (throw (ex-info "No such workload" {:uuid uuid})))
+              (postgres/load-workloads tx))))))))
 
 (defn post-start
   "Start the workloads with UUIDs in REQUEST."
   [request]
-  (wfl.api.handlers/print-handler-error
-   (let [uuids (-> request :parameters :body distinct)]
-     (logr/infof "post-start endpoint called: uuids=%s" uuids)
-     (letfn [(q [[left right]] (fn [it] (str left it right)))]
-       (let [db-config (postgres/wfl-db-config)
-             db-conn (jdbc/get-connection db-config)
-             query (->> (repeat (count uuids) "?")
-                        (str/join ",") ((q "()"))
-                        (format "SELECT * FROM workload WHERE uuid in %s"))
-             ps (jdbc/prepare-statement db-conn query)]
-         (doseq [[i uuid] (map-indexed vector uuids)] (.setString ps (+ i 1) (:uuid uuid)))
-         (jdbc/with-db-transaction [tx db-config]
-           (->> ps
-                (jdbc/query tx)
-                (run! (partial start-workload! tx)))
-           (->> uuids
-                (mapv (partial postgres/get-workload-for-uuid tx))
-                succeed)))))))
-
-(def post-exec
+  (letfn [(go! [tx {:keys [uuid]}]
+            (if-let [workload (postgres/load-workload-for-uuid tx uuid)]
+              (do
+                (start-workload! tx workload)
+                (postgres/load-workload-for-id tx (:id workload)))
+              (throw
+                (ex-info "No such workload" {:uuid uuid}))))]
+    (log-and-fail-on-error
+      (jdbc/with-db-transaction [tx (postgres/wfl-db-config)]
+        (let [uuids (-> request :parameters :body distinct)]
+          (logr/infof "post-start endpoint called: uuids=%s" uuids)
+          (succeed (mapv #(go! tx %) uuids)))))))
+
+(defn post-exec
   "Create and start workload described in BODY of REQUEST"
-  (let [pack   (fn [response] {:parameters {:body [(:body response)]}})
-        unpack (comp first :body)]
-    (comp succeed unpack post-start pack post-create)))
+  [request]
+  (letfn [(create! [tx workload-request]
+            (->>
+              (add-workload! tx workload-request)
+              :uuid
+              (postgres/load-workload-for-uuid tx)))
```

Comment:
I think add-workload should return the created workload rather than just its UUID.

Reply:
Works for me.